### PR TITLE
DBZ-9557 Add spanner change stream connector support for placement tables.

### LINKED ISSUE #1415: Add spanner change stream connector support for placement tables. [DBZ-9557]
Migrated from [DBZ-9557](https://issues.redhat.com/browse/DBZ-9557)

Add spanner change stream connector support for placement tables.

### PR DESCRIPTION
**Pull request by g-zemlianyi**
Closes debezium/dbz#1415

### CODE REVIEW

**Comment by Naros** [comment]
Hi @g-zemlianyi, please be sure to follow the contributing guidelines, sign your comments, and ensure the commit messages are accurately prefixed. It seems this is related to debezium/dbz#1415; however, that issue has since been closed. If it is still related, please use that issue key. Thanks.

**Review by jiangzzhu — APPROVED** [review]
LGTM. The review process is done by merging code into improving/dev branch.

**Comment by g-zemlianyi** [comment]
Hi @Naros, thank you for the suggestions. I’ve fixed the commit names and signed them. No further changes are expected in this branch, so the PR is ready for review.

**Comment by jiangzzhu** [comment]
@Naros and @jpechane, could you help us review and merge the PR? Thanks!

File: src/main/java/io/debezium/connector/spanner/db/stream/SpannerChangeStreamService.java [hunk 1]
```diff
@@ -97,10 +134,173 @@ public void getEvents(Partition partition, ChangeStreamEventConsumer changeStrea
...
+                    List<ChangeStreamEvent> events = filterBoundaryDuplicates(rawEvents, processedTimestamp, lastBoundaryRecordSequence);
+
+                    if (!events.isEmpty() && (events.get(0) instanceof HeartbeatEvent)) {
+                        var heartbeatEvent = (HeartbeatEvent) events.get(0);
+                        long heartbeatLag = System.currentTimeMillis() - heartbeatEvent.getRecordTimestamp().toSqlTimestamp().toInstant().toEpochMilli();
+                        if (heartbeatLag > 60_000) {
+                            LOGGER.warn("Task: {}, heartbeat has very old timestamp, lag: {}, token: {}, event: {}", taskUid, heartbeatLag,
+                                    heartbeatEvent.getMetadata().getPartitionToken(),
+                                    heartbeatEvent);
+                        }
```
Comment (Naros) [inline src/main/java/io/debezium/connector/spanner/db/stream/SpannerChangeStreamService.java]: Is there a possibility that the 60-second threshold should be configurable, even if it's an internal Field using `Field.createInternal` for those corner cases where it may?

**Review by Naros — COMMENTED** [review]
Overall LGTM, just one inine comment below.

File: src/main/java/io/debezium/connector/spanner/db/stream/SpannerChangeStreamService.java (hunk 1 above)
Comment (g-zemlianyi) [reply src/main/java/io/debezium/connector/spanner/db/stream/SpannerChangeStreamService.java]: Hi @Naros 

Sorry for the late response, and thank you for the feedback. I’ve moved the property to the config as you requested.

I also wanted to ask about some additional changes that came up last week. We were quite busy running load testing with the placement tables, moving data across geo-partitions, and testing the connector under heavy load with a high frequency of partition move-in/move-out events (around every ~100 ms).

The testing showed that the existing architecture needed to be revisited, particularly around transitioning through the state machine and handling the destination partition on a move-in event. Based on that, I implemented a rail switch to buffer events until the corresponding move-in/move-out pair is resolved, and then continue flushing the buffered events.

The changes are around 2K lines and are currently under review by @jiangzzhu.

Could you please advise what would be the better approach from a review/PR perspective: should I commit these changes to this PR, or would it be better to create a new PR for the same GitHub issue?

Thanks!

Comment (jiangzzhu) [reply src/main/java/io/debezium/connector/spanner/db/stream/SpannerChangeStreamService.java]: Hi @Naros and @jpechane , do you have any suggestion on above issue? Thanks!

Comment (g-zemlianyi) [reply src/main/java/io/debezium/connector/spanner/db/stream/SpannerChangeStreamService.java]: I added one more commit to resolve the merge conflict.

**Comment by Naros** [comment]
Hi @g-zemlianyi, could you do us a favor? Can you remove the merge commit and use rebase?

File: src/main/java/io/debezium/connector/spanner/db/stream/SpannerChangeStreamService.java (hunk 1 above)
Comment (Naros) [reply src/main/java/io/debezium/connector/spanner/db/stream/SpannerChangeStreamService.java]: Hi @g-zemlianyi, great question. 

For community connectors, we generally defer to the community teams on what they think is best, since you are far more aware of a feature's nuances than we often are. 

My personal rule of thumb is: if a new feature lands across two PRs and only one makes it into a final release, does that leave the connector knowingly unstable? If so, I typically bundle the code into one PR. If not, a follow-up is good to keep things moving. But that's just a personal PoV.

But also keep in mind that you can always open a second PR based on this PR and just highlight in the PR description that they're stacked on one another so we understand the dependency, and when reviewing the second PR, we focus on the specific commits that come at the end of that history trail.

So it's whatever you believe is best for you, the team, and the connector's stability.

**Comment by g-zemlianyi** [comment]
Hi @Naros , thanks for the comment! I rebased on main and force-pushed the updated commits. I also added a bunch of new commits, including performance optimizations and Postgres dialect support (FYI @jiangzzhu).

**Review by jiangzzhu — APPROVED** [review]

**Comment by jiangzzhu** [comment]
Hi @Naros and @jpechane , could you help us review and merge the PR when you have time? Thanks!

**Comment by jiangzzhu** [comment]
Friendly ping.

**Comment by g-zemlianyi** [comment]
Hi @Naros , would it be possible for you to review this PR when you have a chance? Thanks!

**Review by Naros — APPROVED** [review]

**Comment by Naros** [comment]
Hi @g-zemlianyi reviewed, and applied.  Thanks :tada:

**Comment by g-zemlianyi** [comment]
Thank you, @Naros. I saw that 3 edge-case IT tests failed out of the 140 new IT tests. I’m going to create a new PR with a new feature for per-placement TVFs and add a new profile to the POM to exclude those tests from the normal run.
